src/Mono.Android/Microsoft.Android.Runtime/ManagedValueManager.cs

```diff
@@ -250,20 +271,163 @@ public override List<JniSurfacedPeerInfo> GetSurfacedPeers ()
+					throw new InvalidOperationException ("Unknown reference tracking handle.");
+				}
+
+				return handle;
+			}
+		}
+
+		public static HandleContext* Alloc (IJavaPeerable peer)
+		{
+			var context = (HandleContext*)NativeMemory.AllocZeroed (1, Size);
+			if (context == null) {
+				throw new OutOfMemoryException ("Failed to allocate memory for HandleContext.");
+			}
+
+			context->identityHashCode = peer.JniIdentityHashCode;
+			context->controlBlock = peer.JniObjectReferenceControlBlock;
+
+			GCHandle handle = JavaMarshal.CreateReferenceTrackingHandle (peer, context);
+			lock (referenceTrackingHandles) {
+				referenceTrackingHandles [(IntPtr)context] = handle;
+			}
+
+			return context;
+		}
+
+		public static void Free (ref HandleContext* context)
+		{
+			if (context != null) {
+				NativeMemory.Free (context);
+				context = null;
+			}
+		}
+	}
+
+	[UnmanagedCallersOnly]
+	static void BridgeProcessingStarted ()
+	{
+		bridgeProcessingSemaphore.Wait ();
+	}
+
+	[UnmanagedCallersOnly]
+	static unsafe void BridgeProcessingFinished (MarkCrossReferencesArgs* mcr)
+	{
+		Trace.Assert (mcr != null, "MarkCrossReferenceArgs should never be null.");
+
+		ReadOnlySpan<GCHandle> handlesToFree = ProcessCollectedContexts (mcr);
+		JavaMarshal.FinishCrossReferenceProcessing (mcr, handlesToFree);
+
+		bridgeProcessingSemaphore.Release ();
+	}
+
+	static unsafe ReadOnlySpan<GCHandle> ProcessCollectedContexts (MarkCrossReferencesArgs* mcr)
+	{
+		List<GCHandle> handlesToFree = [];
+		ManagedValueManager instance = GetOrCreateInstance ();
+
+		for (int i = 0; (nuint)i < mcr->ComponentCount; i++) {
+			StronglyConnectedComponent component = mcr->Components [i];
+			for (int j = 0; (nuint)j < component.Count; j++) {
+				ProcessContext ((HandleContext*)component.Contexts [j]);
+			}
+		}
+
+		void ProcessContext (HandleContext* context)
+		{
+			Trace.Assert (context != null, "Context should never be null.");
+
+			// Ignore contexts which were not collected
+			if (!context->IsCollected) {
+				return;
+			}
+
+			GCHandle handle = HandleContext.GetAssociatedGCHandle (context);
+
+			// Note: modifying the RegisteredInstances dictionary while processing the collected contexts
+			// is tricky and can lead to deadlocks, so we remember which contexts were collected and we will free
+			// them later outside of the bridge processing loop.
+			instance.CollectedContexts.Enqueue ((IntPtr)context);
+
+			// important: we must not free the handle before passing it to JavaMarshal.FinishCrossReferenceProcessing
+			handlesToFree.Add (handle);
+		}
+
+		return CollectionsMarshal.AsSpan (handlesToFree);
+	}
+
+	const BindingFlags ActivationConstructorBindingFlags = BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance;
 
-	static  readonly    Type[]  XAConstructorSignature  = new Type [] { typeof (IntPtr), typeof (JniHandleOwnership) };
+	static readonly Type[] XAConstructorSignature = new Type[] { typeof (IntPtr), typeof (JniHandleOwnership) };
```

Comment:
nit: maybe we could just put this back to make the diff smaller:
```suggestion
	static  readonly    Type[]  XAConstructorSignature  = new Type [] { typeof (IntPtr), typeof (JniHandleOwnership) };
```

Jon Pryor had some editor that padded with spaces. We don't have to use it, but maybe we can keep old code the way it was.

The padding can make diffs easier to read when adding new C# keywords or changing types, example:
```diff
--		public          JniObjectReference          PeerReference {
++		public  unsafe  JniObjectReference          PeerReference {
```

Reply:
I've been struggling with VS Code a lot while working on this PR. It really doesn't like extra white space and "egyptian brackets" :D I will try to revisit this next week.

Reply:
Yes, something keeps auto-formatting in VS Code, I haven't figured it out yet either...

Best I found is to press Ctrl+Z (Cmd+Z) when I notice it.